File: solutions/typescript/pythagorean-triplet/1/pythagorean-triplet.ts
type Options = {
  minFactor?: number
  maxFactor?: number
  sum: number
}

export function triplets({ minFactor = 1, maxFactor, sum }: Options): Triplet[] {
  const results: Triplet[] = [];
  
  // If maxFactor is not provided, calculate a reasonable upper bound
  // Since a < b < c and a + b + c = sum, the maximum possible value for a is sum/3
  const maxA = Math.floor(sum / 3);
  
  // Determine the search range
  const startA = Math.max(minFactor, 1);
  const endA = maxFactor ? Math.min(maxFactor, maxA) : maxA;
  
  for (let a = startA; a <= endA; a++) {
    // For a given a, b must be greater than a and less than (sum - a)/2
    const maxB = Math.floor((sum - a) / 2);
    const startB = a + 1;
    const endB = maxFactor ? Math.min(maxFactor, maxB) : maxB;
    
    for (let b = startB; b <= endB; b++) {
      // Calculate c from the sum constraint
      const c = sum - a - b;
      
      // Check if c satisfies the triplet conditions
      if (c <= b) continue; // c must be greater than b
      if (maxFactor && c > maxFactor) continue; // Check maxFactor constraint
      
      // Check Pythagorean theorem
      if (a * a + b * b === c * c) {
        results.push(new Triplet(a, b, c));
      }
    }
  }
  
  return results;
}

class Triplet {
  private a: number;
  private b: number;
  private c: number;
  
  constructor(a: number, b: number, c: number) {
    this.a = a;
    this.b = b;
    this.c = c;
  }
  
  toArray(): [number, number, number] {
    return [this.a, this.b, this.c];
  }
}

File: solutions/ruby/pythagorean-triplet/1/pythagorean_triplet.rb
class PythagoreanTriplet
  def self.triplets_with_sum(sum)
    triplets = []
    
    # a must be less than sum/3 since a < b < c
    (1...sum / 3).each do |a|
      # From a² + b² = c² and a + b + c = sum:
      # c = sum - a - b
      # Substituting and solving for b:
      # b = sum(sum - 2a) / 2(sum - a)
      
      numerator = sum * (sum - 2 * a)
      denominator = 2 * (sum - a)
      
      next if numerator % denominator != 0
      
      b = numerator / denominator
      c = sum - a - b
      
      # Validate: a < b < c and Pythagorean theorem holds
      if b > a && c > b && a**2 + b**2 == c**2
        triplets << [a, b, c]
      end
    end
    
    triplets
  end
end